py3status/modules/volume_status.py

```diff
@@ -223,8 +224,14 @@ def get_default_device(self):
 
     def get_volume(self):
         output = self.command_output(['pactl', 'list', self.device_type_pl]).strip()
-        muted, perc = self.re_volume.search(output).groups()
 
+        output_matches = self.re_volume.search(output)
+        if output_matches is None:
+            # This happens when device changed. Try again.
+            self.device = None
```

Comment:
self.device for sure - see the check in line 193 (if self.device is None: ). Probably cleaner to change both. Looking at @lasers comment and #924 , I don't handle the case when user has explicitly specified a device. It seems that not specifying a device should mean 'auto' and allow for switching automatically (current behavior in this commit). In case they do specify explicitly, it should not switch (need to add this check). @lasers , lmk if this makes sense and I'll submit another one soon. Thanks!

Reply:
>self.device for sure

`self.setup(self.parent)` right after `self.device = None` is why I asked.

>In case they do specify explicitly, it should not switch (need to add this check). @lasers , lmk if this makes sense and I'll submit another one soon. Thanks!

Yes, it would make sense. I'm going to wait for @ultrabug first to decide what to do here. I optimized my old PR just now (nothing significant changed). I favor my PR more because it `try` instead of checking first and it does not have to `setup()` variables, just `device`.

Also, changing sink sources can be very infrequent so it would also make sense to use my PR. Maybe you could use my PR as yours since I couldn't get this merged in first place without proof of improvements. I _just_ didn't have a star witness before. :-)

Reply:
I've updated my diff and @ultrabug can choose. This diff calls setup() again to account for more issues/state changes but who really knows if that's better or worse. Don't feel strongly either way. Thanks!

Reply:
ok maybe not. :-)

@lasers, Just to check, I copied https://raw.githubusercontent.com/lasers/py3status/volume-status-fix-pactl/py3status/modules/volume_status.py and the issue is when I use gnome-control-center to change output device, it is still showing the output volume of the old device. The issue seems to be line 239 - it just picks the first device which in my case is suspended (I have multiple audio devices). Running get_default_device/"pactl info" is probably more robust.

Reply:
Cycles through 0 for 1 second and then updates with the right value. Why not get output again after catching AttributeError? Is this to show 0 when there are no devices? If so, I guess its an ok tradeoff.

Reply:
It can be better by updating with the right values too. I feel this is a nice "reset device" feedback for users to see (only if they switched sinks very recently). Most of the time, users won't see this because they don't switch sinks enough or that next interval would be corrected not long after (either within 10 seconds or when users started scrolling/clicking).

Reply:
Agreed. It fixes the core issue of having to restart upon device change while not redoing all the setup() stuff. Works for me. Thank you. I'll close this one.

Reply:
Also, if the `device` was specified and is not available, the user would see `0` too. I will put out a PR. Thank you for working out the imperfections.